tempodb/encoding/backend_block.go

```diff
@@ -0,0 +1,122 @@
+package encoding
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/opentracing/opentracing-go"
+	willf_bloom "github.com/willf/bloom"
+
+	"github.com/grafana/tempo/tempodb/backend"
+	"github.com/grafana/tempo/tempodb/encoding/common"
+)
+
+// BackendBlock represents a block already in the backend.
+type BackendBlock struct {
+	encoding versionedEncoding
+
+	meta   *backend.BlockMeta
+	reader backend.Reader
+}
+
+// NewBackendBlock returns a BackendBlock for the given backend.BlockMeta
+//  It is version aware.
+func NewBackendBlock(meta *backend.BlockMeta, r backend.Reader) (*BackendBlock, error) {
+	var encoding versionedEncoding
+
+	switch meta.Version { // jpe need a way to test this
+	case "v0":
+		encoding = v0Encoding{}
+	case "v1":
+		encoding = v1Encoding{}
+	default:
+		return nil, fmt.Errorf("%s is not a valid block version", meta.Version)
+	}
+
+	return &BackendBlock{
+		encoding: encoding,
+		meta:     meta,
+		reader:   r,
+	}, nil
+}
+
+// Find searches a block for the ID and returns an object if found.
+func (b *BackendBlock) Find(ctx context.Context, id common.ID) ([]byte, error) {
+	var err error
+	span, ctx := opentracing.StartSpanFromContext(ctx, "BackendBlock.Find")
+	defer func() {
+		if err != nil {
+			span.SetTag("error", true)
+		}
+		span.Finish()
+	}()
+
+	span.SetTag("block", b.meta.BlockID.String())
+
+	shardKey := common.ShardKeyForTraceID(id)
+	blockID := b.meta.BlockID
+	tenantID := b.meta.TenantID
+
+	bloomBytes, err := b.reader.Read(ctx, b.encoding.nameBloom(shardKey), blockID, tenantID)
```

Comment:
Using this spot as a discussion point in general:   This might not be in the direction we were going with the recent refactors.  The presence of bloom filters, for example, is something that could/should be behind versioning.  In this case v0/v1 are so similar it makes sense to combine both here for code reuse, but wonder if there should be separate BackendBlocks in v0 and v1.  Or rename/relocate the logic here to indicate that it is a BackendBlock with both v0/v1 compatibility.

Reply:
As I moved through the refactor I really started liking the backend, compaction and complete blocks all using the encoding interface to interact with the data in the backend.  The `versionedEncoding` interface is a bit gross at the moment, but I didn't want to take the refactor any farther for obvious reasons.  This PR is already quite beefy.  

Regardless, I think `versionedEncoding` will be a sufficient shim in the future to handle changes to the way a trace is found.